db/blip_handler.go

```diff
@@ -1083,41 +1099,70 @@ func (bsc *BlipSyncContext) incrementSerialNumber() uint64 {
 	return atomic.AddUint64(&bsc.handlerSerialNumber, 1)
 }
 
-func (bsc *BlipSyncContext) addAllowedAttachments(attDigests []string) {
-	if len(attDigests) == 0 {
+func (bsc *BlipSyncContext) addAllowedAttachments(docID string, attMeta []AttachmentStorageMeta) {
+	if len(attMeta) == 0 {
 		return
 	}
 
 	bsc.lock.Lock()
 	defer bsc.lock.Unlock()
 
 	if bsc.allowedAttachments == nil {
-		bsc.allowedAttachments = make(map[string]int, 100)
-	}
-	for _, digest := range attDigests {
-		bsc.allowedAttachments[digest] = bsc.allowedAttachments[digest] + 1
+		bsc.allowedAttachments = make(map[string]AllowedAttachment, 100)
+	}
+	for _, attachment := range attMeta {
+		key := bsc.allowedAttachmentKey(docID, attachment.digest)
+		att, found := bsc.allowedAttachments[key]
+		if found {
+			if att.version == AttVersion1 {
+				att.counter = att.counter + 1
+				bsc.allowedAttachments[key] = att
+			} else if att.version == AttVersion2 {
+				bsc.allowedAttachments[key] = AllowedAttachment{}
```

Comment:
Why are you setting the value to an empty AllowedAttachment{} -  won't this lose the docID needed to retrieve the attachment?  Please ensure there is a unit test covering this path.

Reply:
That was not logically correct. We will miss the docID to retrieve the attachment if we set the value to an empty AllowedAttachment{}. The else block can be removed since we don't need to increment or decrement the counter for AttVersion2. And the value of the counter will be set to 1 when the attachment is added to allowedAttachments map in the first place. TestPutAttachmentViaBlipGetViaBlip has the coverage for V1 and V2 attachments handling. Also added a change in storeAttachments set the version if it's passed the document body along with attachment metadata. It's necessary for adding the coverage.